File: Sigmath/Sigmath/Sigmath.ConvertedToVBNet/computing/sets.vb
Imports System.Collections.Generic
Imports System.Linq

Public Class Sets
	Public Shared Function Union(Of T)(set1 As List(Of T), set2 As List(Of T)) As List(Of T)
		Dim unionSet = New List(Of T)(set1)
		unionSet.AddRange(set2.Except(set1))
		Return unionSet
	End Function

	Public Shared Function Intersection(Of T)(set1 As List(Of T), set2 As List(Of T)) As List(Of T)
		Dim intersectionSet = New List(Of T)(set1)
		intersectionSet.RemoveAll(Function(x) Not set2.Contains(x))
		Return intersectionSet
	End Function

	Public Shared Function Difference(Of T)(set1 As List(Of T), set2 As List(Of T)) As List(Of T)
		Dim differenceSet = New List(Of T)(set1)
		differenceSet.RemoveAll(Function(x) set2.Contains(x))
		Return differenceSet
	End Function

	Public Shared Function SymmetricDifference(Of T)(set1 As List(Of T), set2 As List(Of T)) As List(Of T)
		Dim symmetricDifferenceSet = New List(Of T)()
		Dim unionSet = Union(set1, set2)

		For Each item As var In unionSet
			If (set1.Contains(item) AndAlso Not set2.Contains(item)) OrElse (Not set1.Contains(item) AndAlso set2.Contains(item)) Then
				symmetricDifferenceSet.Add(item)
			End If
		Next

		Return symmetricDifferenceSet
	End Function

	Public Shared Function IsSubset(Of T)(subset As List(Of T), [set] As List(Of T)) As Boolean
		Return subset.All(AddressOf [set].Contains)
	End Function

	Public Shared Function IsSuperset(Of T)(superset As List(Of T), [set] As List(Of T)) As Boolean
		Return [set].All(AddressOf superset.Contains)
	End Function
End Class

File: computing/sets.cs
using System.Collections.Generic;
using System.Linq;

namespace Mathian
{
    public class Sets
    {
        public static List<T> Union<T>(List<T> set1, List<T> set2)
        {
            var unionSet = new List<T>(set1);
            unionSet.AddRange(set2.Except(set1));
            return unionSet;
        }

        public static List<T> Intersection<T>(List<T> set1, List<T> set2)
        {
            var intersectionSet = new List<T>(set1);
            intersectionSet.RemoveAll(x => !set2.Contains(x));
            return intersectionSet;
        }

        public static List<T> Difference<T>(List<T> set1, List<T> set2)
        {
            var differenceSet = new List<T>(set1);
            differenceSet.RemoveAll(x => set2.Contains(x));
            return differenceSet;
        }

        public static List<T> SymmetricDifference<T>(List<T> set1, List<T> set2)
        {
            var symmetricDifferenceSet = new List<T>();
            var unionSet = Union(set1, set2);

            foreach (var item in unionSet)
            {
                if ((set1.Contains(item) && !set2.Contains(item))
                    || (!set1.Contains(item) && set2.Contains(item)))
                {
                    symmetricDifferenceSet.Add(item);
                }
            }

            return symmetricDifferenceSet;
        }

        public static bool IsSubset<T>(List<T> subset, List<T> set)
        {
            return subset.All(set.Contains);
        }

        public static bool IsSuperset<T>(List<T> superset, List<T> set)
        {
            return set.All(superset.Contains);
        }
    }
}